src/interpreter/vm.rs

```diff
@@ -63,6 +67,112 @@ pub struct BlockMessages {
     pub win_count: i64,
 }
 
+#[derive(PartialEq, Clone, Debug)]
+pub struct MessageGasCost {
+    pub message: Option<Cid>,
+    pub gas_used: BigInt,
+    pub base_fee_burn: TokenAmount,
+    pub over_estimation_burn: TokenAmount,
+    pub miner_penalty: TokenAmount,
+    pub miner_tip: TokenAmount,
+    pub refund: TokenAmount,
+    pub total_cost: TokenAmount,
+}
+
+impl MessageGasCost {
+    pub fn new(msg: &Message, ret: &ApplyRet) -> Self {
+        use crate::message::Message as MessageTrait;
+        Self {
+            message: Some(msg.cid().unwrap()),
+            gas_used: BigInt::from(ret.msg_receipt().gas_used()),
+            base_fee_burn: ret.base_fee_burn(),
+            over_estimation_burn: ret.over_estimation_burn(),
+            miner_penalty: ret.penalty(),
+            miner_tip: ret.miner_tip(),
+            refund: ret.refund(),
+            total_cost: msg.required_funds() - &ret.refund(),
+        }
+    }
+
+    pub fn from_implicit(msg: &Message, ret: &ApplyRet) -> Self {
+        Self {
+            message: None,
+            gas_used: BigInt::default(),
+            ..Self::new(msg, ret)
+        }
+    }
+}
+
+#[derive(PartialEq, Clone, Debug)]
+pub struct InvocResult {
+    pub msg_cid: Cid,
+    pub msg: Message,
+    pub msg_receipt: Receipt,
+    pub gas_cost: MessageGasCost,
+    pub execution_trace: Option<Trace>,
+    pub error: String,
+    pub duration: u64,
+}
+
+impl InvocResult {
+    pub fn new(msg_cid: Cid, msg: &Message, ret: &ApplyRet) -> Self {
+        let trace = build_exec_trace(ret.exec_trace());
+        Self {
+            msg_cid,
+            msg: msg.clone(),
+            msg_receipt: ret.msg_receipt(),
+            gas_cost: MessageGasCost::new(msg, ret),
+            execution_trace: trace,
+            error: ret.failure_info().unwrap_or_default(),
+            duration: 0,
+        }
+    }
+
+    pub fn from_implicit(msg_cid: Cid, msg: &Message, ret: &ApplyRet) -> Self {
+        Self {
+            gas_cost: MessageGasCost::from_implicit(msg, ret),
+            ..Self::new(msg_cid, msg, ret)
+        }
+    }
+}
+
+fn build_exec_trace(exec_trace: Vec<ExecutionEvent_v3>) -> Option<Trace> {
+    let exec_trace: Option<Trace> = if !exec_trace.is_empty() {
+        let mut trace_iter = exec_trace.into_iter();
+        let mut initial_gas_charges = Vec::new();
+        loop {
+            match trace_iter.next() {
```

Comment:
How about `while let Some(..) = trace_iter.next()` ?

Reply:
I don't think it's possible because you can only break with a value inside a `loop` construct.